### PR TITLE
feat(1415): Add alternate routes and handlers for subscribed repos

### LINKED ISSUE #1415: Notify SCM changes from external repositories
## Context

There are many pipelines whose builds are dependent on source changes happening across  external SCM repositories in addition to repository from which it was created.
For example `screwdriver/screwdriver-release` pipeline which needs to be built whenever a new `tag` or `release` on `screwdriver/screwdriver` or `screwdriver/ui` repositories.

## Objective

1. Add webhooks to subscribed pipelines and when webhook is triggered it should match the subscriber pipelines which are subscribed to it.
    1. If user does not have permission to add webhooks on any of the subscribed repositories, display appropriate message guiding the user with manual addition of the webhooks. 
1. SCM change in the subscribed repository should be captured in  `metadata` of the build so that subscriber pipelines can do necessary code checkouts or build commands based on change
1. It is the responsibility of subscriber pipeline to do the necessary code checkouts. [TBD]
    1.  Git Submodules can be used for this without having SD implement the logic

### Config

```

subscribe:
   scmUrls:
      - git@github.com:minz1027/test.template.git: [~commit, ~tags, ~release]
      - git@github.com:minz1027/quickstart-generic.git#master: [~commit, ~tags, ~release]
jobs:
  main:
    requires: [~subscribe, ~commit]
```

## References

1. Trigger on release/tags - https://github.com/screwdriver-cd/screwdriver/issues/823

### PR DESCRIPTION
**Pull request by supra08**
## Context

Fixes screwdriver-cd/screwdriver#1415

## Objective

This PR adds handlers for using the subscribed config. If the parsed config contains the `subscribe` object it is formatted to convert scmUrl to scmUri mapped with their actions. This mapped array is used by the modified addWebhook() method (now addMultipleWebhooks()) to add the webhooks along with the corresponding actions.

Now when a webhook request arrives in the screwdriver API, it identifies it whether it is external and captures is `checkoutUrl`. The API filters the triggered pipelines based on the `checkoutUrl`. Further in the `eventFactory`, it needs to be figured whether the job(s) is(are) triggered or not. This is done by performing a `breadth-first-search` of the `workflowGraph` with `~subscribe` as the root node. This way all the dependant jobs are identified wrt the triggered jobs, `workflowGraph` updated and build started.

## References

screwdriver-cd/screwdriver#1415

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: lib/eventFactory.js [hunk 1]
```diff
@@ -108,6 +108,9 @@ function getJobsFromPR(config) {
     const { jobs, prNum, pipelineConfig, startFrom, chainPR, branch } = config;
     const isPRTrigger = PR_TRIGGER.test(startFrom);
 
+    // console.log('here - pr - 1000');
+    // console.log(jobs);
+
```
Comment (pritamstyz4ever) [inline lib/eventFactory.js]: Remove commented code

File: lib/eventFactory.js [hunk 2]
```diff
@@ -133,6 +136,92 @@ function getJobsFromPR(config) {
...
+Queue.prototype.enqueue = e => {
+    this.elements.push(e);
+};
+
+Queue.prototype.isEmpty = () => {
+    return this.elements.length === 0;
+};
+
+Queue.prototype.dequeue = () => {
```
Comment (pritamstyz4ever) [inline lib/eventFactory.js]: Do you mean this.elements.shift() here?

Comment (pritamstyz4ever) [inline lib/eventFactory.js]: Consider converting to class structure.

File: lib/eventFactory.js [hunk 3]
```diff
@@ -133,6 +136,92 @@ function getJobsFromPR(config) {
...
+ * Filter workflow graph by bfs traversal wrt the subscribed job
+ * @method filterSubscribedJobs
+ * @param  {Object}    workflowGraph
+ * @return {Object}                                         Filtered workflow graph
+ */
+function filterSubscribedJobs(workflowGraph) {
+    const adj = {};
+    const reachableNodes = [];
+    const visitedNode = {};
+    const queue = new Queue();
```
Comment (pritamstyz4ever) [inline lib/eventFactory.js]: Can we write the same logic without creating a class Queue and using an array instead. 
`const queue = [];`

**Review by pritamstyz4ever — COMMENTED** [review]
Some comments

File: lib/eventFactory.js (hunk 3 above)
Comment (supra08) [reply lib/eventFactory.js]: Yeah, that could be done. I was thinking of standard DS usage but that's not necessary here I guess.

File: lib/eventFactory.js (hunk 2 above)
Comment (supra08) [reply lib/eventFactory.js]: I think better use the simple array only. eslint errored out on using another class beside EventFactory

File: lib/pipeline.js [hunk 4]
```diff
@@ -747,6 +793,37 @@ class PipelineModel extends BaseModel {
         this.annotations = parsedConfig.annotations;
         this.parameters = parsedConfig.parameters;
 
+        if ('subscribe' in parsedConfig) {
+            if (Object.keys(parsedConfig.subscribe).length !== 0) {
+                const urlWithActionsList = [];
+
+                parsedConfig.subscribe.scmUrls.forEach(url => {
```
Comment (jithine) [inline lib/pipeline.js]: better to refactor [following](https://github.com/screwdriver-cd/models/pull/472/files#diff-679e90b5b15ab1b73ddb7ed2377e7413R800-R809) like this

```
const urlWithActionsList = [];
for (let i = 0; i < parsedConfig.subscribe.scmUrls.length; ++i) {
  let urlObj = parsedConfig.subscribe.scmUrls[i];
  let scmUrl = Object.keys(urlObj)[0];
  urlWithActionsList.push({
    scmUrl: Object.keys(url)[0],
    actions: urlObj[scmUrl].map(action => action[0] === '~'  ? action.substring(1) : action),
    scmUri:  await this._getScmUri(scmUrl); 
  });
}
parsedConfig.subscribe.scmUrls  = urlWithActionsList;
```

File: lib/pipeline.js [hunk 5]
```diff
@@ -747,6 +793,37 @@ class PipelineModel extends BaseModel {
...
+                    urlWithActionsList.push({
+                        scmUrl: Object.keys(url)[0],
+                        actions: Object.values(url)[0]
+                    });
+                });
+                parsedConfig.subscribe.scmUrls = urlWithActionsList.map(urlObject =>
+                    this._formatUrlsWithActionsStruct(urlObject)
+                );
+                this.subscribedScmUrlsWithActions = await Promise.all(parsedConfig.subscribe.scmUrls);
+                this.subscribedScmUrlsWithActions.map(urlWithActions => {
```
Comment (jithine) [inline lib/pipeline.js]: this `map` is  not needed, covered in my previous comment.

File: lib/pipeline.js [hunk 6]
```diff
@@ -373,22 +374,52 @@ class PipelineModel extends BaseModel {
...
     /**
-     * Attach Screwdriver webhook to the pipeline's repository
+     * Attach Screwdriver webhooks to the pipeline's repository
+     * @method addMultipleWebhooks
      * @param   {String}    webhookUrl    The webhook to be added
      * @method  addWebhook
      * @return  {Promise}
      */
-    addWebhook(webhookUrl) {
+    async addMultipleWebhooks(webhookUrl) {
```
Comment (jithine) [inline lib/pipeline.js]: let's just say `addWebhooks`

File: lib/pipeline.js [hunk 7]
```diff
@@ -373,22 +374,52 @@ class PipelineModel extends BaseModel {
...
+        const webhookUrlWithActions = {
+            webhookUrl,
+            actions: ['push', 'pull_request', 'create', 'release'],
+            scmUri: this.scmUri
+        };
+
+        webhookUrlsWithActionsList.push(webhookUrlWithActions);
+
+        // check if subscribedScmUrlsWithActions is present
+        if ('subscribedScmUrlsWithActions' in this && this.subscribedScmUrlsWithActions !== undefined) {
```
Comment (jithine) [inline lib/pipeline.js]: can this be replaced with `if (Array.isArray(this.subscribedScmUrlsWithActions)) {`

File: lib/eventFactory.js [hunk 8]
```diff
@@ -640,6 +737,10 @@ class EventFactory extends BaseFactory {
                 .then(p => {
                     if (prevChainPR !== p.chainPR) {
```
Comment (jithine) [inline lib/eventFactory.js]: just change this to `if (prevChainPR !== p.chainPR && !subscribedEvent)`

File: lib/eventFactory.js [hunk 9]
```diff
@@ -667,7 +768,12 @@ class EventFactory extends BaseFactory {
                     if (prRef) {
                         modelConfig.pr.ref = prRef;
 
-                        return p.syncPR(prNum);
+                        // return p.syncPR(prNum);
+                        if (subscribedEvent) {
+                            return p;
+                        }
+
+                        return p.syncPRs();
```
Comment (jithine) [inline lib/eventFactory.js]: wht did this change to `syncPRs` ?

File: lib/eventFactory.js [hunk 10]
```diff
@@ -686,17 +792,29 @@ class EventFactory extends BaseFactory {
...
 
+                                        let adaptedScmUri = pipeline.scmUri;
+
+                                        if (subscribedEvent) {
+                                            adaptedScmUri = subscribedScmConfig.scmUri;
+                                        }
+
                                         return this.scm.decorateCommit({
-                                            scmUri: pipeline.scmUri,
+                                            scmUri: adaptedScmUri,
```
Comment (jithine) [inline lib/eventFactory.js]: why is this change needed ? decoration is still against current pipeline's `scm` commit right ?

File: lib/eventFactory.js [hunk 11]
```diff
@@ -686,17 +792,29 @@ class EventFactory extends BaseFactory {
...
 
+                            let adaptedScmUri = pipeline.scmUri;
+
+                            if (subscribedEvent) {
+                                adaptedScmUri = subscribedScmConfig.scmUri;
+                            }
+
                             return this.scm.decorateCommit({
-                                scmUri: pipeline.scmUri,
+                                scmUri: adaptedScmUri,
```
Comment (jithine) [inline lib/eventFactory.js]: why is this change needed ? decoration is still against current pipeline's scm commit right ?

File: lib/eventFactory.js [hunk 12]
```diff
@@ -455,14 +557,7 @@ function updateEventParameters(config) {
...
         if (typeof val === 'object') {
-            if (Array.isArray(val)) {
-                defaultParameterValue = val[0];
-            } else {
-                defaultParameterValue = val.value;
-                if (Array.isArray(val.value)) {
-                    defaultParameterValue = val.value[0];
-                }
-            }
+            defaultParameterValue = val.value;
```
Comment (jithine) [inline lib/eventFactory.js]: seems like PR is not in sync with latest in `master` branch

File: lib/eventFactory.js [hunk 13]
```diff
@@ -369,6 +464,13 @@ function createBuilds(config) {
 function getLatestWorkflowGraph(config) {
     const { pipeline, eventConfig } = config;
 
+    // Experimental feature turned off
+    if ('subscribedEvent' in config) {
```
Comment (jithine) [inline lib/eventFactory.js]: can't this just be `if (config.subscribedEvent) {`

File: lib/eventFactory.js [hunk 14]
```diff
@@ -133,6 +136,92 @@ function getJobsFromPR(config) {
...
+    return this.items.shift();
+};
+
+/**
+ * Filter workflow graph by bfs traversal wrt the subscribed job
+ * @method filterSubscribedJobs
+ * @param  {Object}    workflowGraph
+ * @return {Object}                                         Filtered workflow graph
+ */
+function filterSubscribedJobs(workflowGraph) {
```
Comment (jithine) [inline lib/eventFactory.js]: why is this needed, won't setting `startFrom` as `subscribed` take care of the logic ?

File: lib/pipeline.js (hunk 4 above)
Comment (supra08) [reply lib/pipeline.js]: I think we also need to resolve the promises for `this._getScmUrl`?

File: lib/eventFactory.js (hunk 9 above)
Comment (supra08) [reply lib/eventFactory.js]: Yeah it was actually syncPRs. But if there is a subscribed event, there is no point in calling the syncPRs because the repository would not be a pipeline in screwdriver and hence should not affect the PR.

File: lib/eventFactory.js (hunk 10 above)
Comment (supra08) [reply lib/eventFactory.js]: Oh I misunderstood this part. Will revert it back.

File: lib/eventFactory.js (hunk 14 above)
Comment (supra08) [reply lib/eventFactory.js]: What I inferred from the problem is that `~subscribed` is an additional flag that acts as a **selector** in addition to any other trigger that should be there as the `startFrom` value.
So in order to remove any ambiguity, I thought of pulling out this selection logic outside.
This can also help in implementing the label based subscription selector as was discussed, more clearly.

File: lib/eventFactory.js [hunk 15]
```diff
@@ -667,7 +674,7 @@ class EventFactory extends BaseFactory {
                     if (prRef) {
                         modelConfig.pr.ref = prRef;
 
-                        return p.syncPR(prNum);
+                        return subscribedEvent ? p : p.syncPR(prNum);
```
Comment (jithine) [inline lib/eventFactory.js:706]: for subscribed event, does `modelConfig.pr` need to change ? What's the implication ?

File: lib/eventFactory.js [hunk 16]
```diff
@@ -687,7 +694,7 @@ class EventFactory extends BaseFactory {
                                         modelConfig.creator = creator;
 
                                         return this.scm.decorateCommit({
-                                            scmUri: pipeline.scmUri,
+                                            scmUri: subscribedEvent ? subscribedScmConfig.scmUri : pipeline.scmUri,
```
Comment (jithine) [inline lib/eventFactory.js]: I think this change is not needed. Does the subscribed repo need to show pipeline repo's build ? Would it be confusing?

File: lib/pipeline.js [hunk 17]
```diff
@@ -596,6 +627,18 @@ class PipelineModel extends BaseModel {
         return scmUri;
     }
 
+    /**
+     * Format scmUrlsWithActions object by converting scmUrl to scmUri
+     * @method _formatUrlsWithActionsStruct
+     * @param  {Object} scmUrlObject  object containing scmUrl and corresponding actions
+     * @return {Promise}
+     */
+    async _formatUrlsWithActionsStruct(scmUrlObject) {
```
Comment (jithine) [inline lib/pipeline.js]: is this called anywhere ?

Comment (supra08) [reply lib/pipeline.js]: Oh forgot to remove it. It's not required.

File: test/lib/pipeline.test.js [hunk 18]
```diff
@@ -1,3 +1,6 @@
+/* eslint-disable max-lines-per-function */
+/* eslint-disable max-statements */
```
Comment (tkyi) [inline test/lib/pipeline.test.js]: should remove?

File: test/lib/pipeline.test.js [hunk 19]
```diff
@@ -611,6 +616,32 @@ describe('Pipeline Model', () => {
             });
         });
 
+        it('subscribed pipelines from config gets added to the model', () => {
```
Comment (tkyi) [inline test/lib/pipeline.test.js]:
```suggestion
        it('adds subscribed pipelines from config to the model', () => {
```

File: lib/eventFactory.js [hunk 20]
```diff
@@ -369,6 +396,11 @@ function createBuilds(config) {
 function getLatestWorkflowGraph(config) {
     const { pipeline, eventConfig } = config;
 
+    // Experimental feature turned off
+    if (config.subscribedEvent) {
```
Comment (tkyi) [inline lib/eventFactory.js:401]: need to update JSdocs for this

File: lib/eventFactory.js [hunk 21]
```diff
@@ -363,12 +390,18 @@ function createBuilds(config) {
  * @param  {Object}         config
  * @param  {Pipeline}       config.pipeline            Pipeline
  * @param  {Object}         config.eventConfig
+ * @param  {Boolean}        config.subscribedEvent
```
Comment (tkyi) [inline lib/eventFactory.js]: missing JSdoc description

**Review by jithine — APPROVED** [review]

**Review by tkyi — APPROVED** [review]

**Review by jithine — APPROVED** [review]
